hashes/src/sha512_256.rs

```diff
@@ -0,0 +1,210 @@
+// Bitcoin Hashes Library
+// Written in 2018 by
+//   Andrew Poelstra <apoelstra@wpsoftware.net>
+//
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+// This module is largely copied from the rust-crypto ripemd.rs file;
+// while rust-crypto is licensed under Apache, that file specifically
+// was written entirely by Andrew Poelstra, who is re-licensing its
+// contents here as CC0.
+
+//! SHA512_256 implementation.
+//!
+//! SHA512/256 is a hash function that uses the sha512 alogrithm but it truncates
+//! the output to 256 bits. It has different initial constants than sha512 so it
+//! produces an entirely different hash compared to sha512. More information at
+//! https://eprint.iacr.org/2010/548.pdf.
+
+use core::str;
+use core::ops::Index;
+use core::slice::SliceIndex;
+
+use crate::{hex, sha512, sha512::BLOCK_SIZE, Error};
+
+/// Engine to compute SHA512/256 hash function.
+///
+/// SHA512/256 is a hash function that uses the sha512 alogrithm but it truncates
+/// the output to 256 bits. It has different initial constants than sha512 so it
+/// produces an entirely different hash compared to sha512. More information at
+/// https://eprint.iacr.org/2010/548.pdf.
+#[derive(Clone)]
+pub struct HashEngine(sha512::HashEngine);
+
+impl Default for HashEngine {
+    fn default() -> Self {
+        HashEngine(sha512::HashEngine {
+            h: [
+                0x22312194fc2bf72c, 0x9f555fa3c84c64c2, 0x2393b86b6f53b151, 0x963877195940eabd,
+                0x96283ee2a88effe3, 0xbe5e1e2553863992, 0x2b0199fc2c85b8aa, 0x0eb72ddc81c52ca2,
```

Comment:
Weirdly, I don't see these values in the linked paper. What's going on?

Reply:
I don't think the proposed initial values were adopted. The standard itself is part of FIPS and the constants that everyone is using is defined in https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.180-4.pdf (section 5.3.6.1)

I linked the paper as I thought it's more useful for why sha512/256 exists in the first place. The FIPS spec doesn't really explain that. I could change the link to the FIPS pdf.

Reply:
That makes sense. Linking both could be also fine. :)